src/main/java/jenkins/plugins/git/traits/CleanAfterCheckoutTrait.java

```diff
@@ -40,8 +41,8 @@
      * Stapler constructor.
      */
     @DataBoundConstructor
-    public CleanAfterCheckoutTrait() {
```

Comment:
Hi @MarkEWaite, how strict do we have to be about backward compatibility for traits? Has to old constructor to be left in place too with a `@Deprecated` annotation and corresponding Javadoc (as it's done in many other classes)?

Reply:
I'll just answer myself. I think the old constructor should also be left in place. Otherwise e.g. existing Job DSL script may fail.

old way
```
multibranchPipelineJob('generated-git-traits-old') {
  branchSources {
    branchSource {
      source {
        git {
          id('1')
          remote('https://github.com/jenkinsci/job-dsl-plugin.git')
          traits {
            cleanBeforeCheckoutTrait() // <--- fails
            cleanAfterCheckoutTrait()  // <--- fails
          }
        }
      }
    }
  }
}
```

new way
```
multibranchPipelineJob('generated-git-traits-new') {
  branchSources { 
    branchSource { 
      source { 
        git { 
          id('2')
          remote('https://github.com/jenkinsci/job-dsl-plugin.git')
          traits { 
            cleanBeforeCheckoutTrait {
              extension {
                deleteUntrackedNestedRepositories(true)
              }
            }
            cleanAfterCheckoutTrait {
              extension {
                deleteUntrackedNestedRepositories(true)
              }
            }
          }
        }
      }
    }
  }
}
```

Reply:
Do i need to deprecate old constructor? Do I need to mark both constructors with `@DataBoundConstructor` or only the new one? I'm not sure how all this stuff works.

Reply:
Only a single constructor is allowed to be annotated with `@DataBoundConstructor`.
Annotate the new one with `@DataBoundConstructor`.
Annotate the old one with `@Deprecated` and add Javadoc with a `@deprecated` description.
For an example see [UserMergeOptions](https://github.com/jenkinsci/git-plugin/blob/git-4.0.0/src/main/java/hudson/plugins/git/UserMergeOptions.java).

Reply:
Okay, updated PR.

Reply:
I just tested the "old way" locally with Job DSL. Unfortunately it still fails. Job DSL seems to always use the `@DataBoundConstructor` annotated constructor. We have to investigate further.

The rest of the PR now looks good to me. Thanks for your effort.

Reply:
I'm not sure what to do here. As I said, my Stapler (or whatever handles `@DataBoundConstructor`) knowledge is close to zero.

Hope that you can come up with a solution, then I'll implement it if such kind of backward-compatibility breakage is unacceptable.

As a last resort measure, we could create a _second_ trait (`cleanAfterCheckoutExtensionTrait` or smth like that), though it feels really awful.